utilities/include/vec4.hpp

```diff
@@ -0,0 +1,157 @@
+            return sum(temp);
+        }
+        
+        CONSTEXPR_CMATH Utype norm() const{
+            return sqrt(dot(*this));
+        }
+
+        CONSTEXPR_CMATH static Utype angle(const Vec4& v1,const Vec4& v2){
+            return acos(v1.dot(v2)/(v1.norm()*v2.norm()));
+        }
+        CONSTEXPR_CMATH Vec4 unit() const{
+            return *this/this->norm();
+        }
+
+        // Negation 
+        friend constexpr Vec4 operator-(const Vec4& v1){
+            return batch2vec(-vec2batch(v1));
+        }
+
+        // Addition and Substraction
+        friend constexpr Vec4& operator+=(Vec4& v1, const Vec4& v2){
+            xsimd::batch<Utype,UArch> b1 = vec2batch(v1);
+            b1+=vec2batch(v2);
+            v1= batch2vec(b1);
+            return v1;
+        }
+
+        friend constexpr Vec4 operator+(const Vec4& v1, const Vec4& v2){
+            Vec4 v1_copy = v1;
+            v1_copy+=v2;
+            return v1_copy;
+        }
+
+        friend constexpr Vec4 operator-(const Vec4& v1, const Vec4& v2){
+            return v1+(-v2);
+        }
+
+        friend constexpr Vec4& operator-=(Vec4& v1, const Vec4& v2){
+            return v1+=-v2;
+        }
+
+        // Scalar Multiplication and Division
+        // scalar post-multiplications
+        friend constexpr Vec4 operator*(const Vec4& v1, const Utype& s){
+            Vec4 v1_copy = v1;
+            v1_copy*=s;
+            return v1_copy;
+        }
+
+        // scalar pre-multiplications
+        friend constexpr Vec4 operator*(const Utype& s, const Vec4& v1){
+            xsimd::batch<Utype,UArch> vs= xsimd::batch<Utype,UArch>::broadcast(s);
+            Vec4 v1_copy = v1;
+            v1_copy*=s;
+            return v1_copy;
+        }
+
+        // scalar post-multiplication-assignment
+        friend constexpr Vec4& operator*=(Vec4& v1, const Utype& s){
+            xsimd::batch<Utype,UArch> vs= xsimd::batch<Utype,UArch>::broadcast(s);
+            v1  = batch2vec(vec2batch(v1)*vs);
+            return v1;
+        }
+        
+        // scalar post-division
+        friend constexpr Vec4 operator/(const Vec4& v1, const Utype& s){
+            return v1*(1/s);
+        }
+
+        // scalar division+assignment
+        friend constexpr Vec4& operator/=(Vec4& v1, const Utype& s){
+            v1*=1/s;
+            return v1;
+        }
+
+        // print the vector
+        friend std::ostream& operator<<(std::ostream& out, const Vec4& v) {            
+            out <<'('<< v.m_v[0] << ',' << v.m_v[1] << ',' << v.m_v[2]<<','<< v.m_v[3]<<')';
+            return out;
+		}
+	
+    protected:
+        static constexpr xsimd::batch<Utype,UArch> vec2batch(const Vec4& v){
+            return  xsimd::batch<Utype,UArch>::load_aligned(v.m_v);
+        }
+        
+        static constexpr Vec4 batch2vec(xsimd::batch<Utype,UArch> x){
+            Vec4 v{x};
+            return v;
+        }
+
+        static constexpr Utype sum(xsimd::batch<Utype,UArch> x){
```

Comment:
This function cannot be labelled `constexpr` because `xsimd::hadd` is not constexpr.

Reply:
I have to make runtime branches for all those functions.